### ISSUE TITLE
RangeError (index): Invalid value: Valid value range is empty: 0 when using if statement

### ISSUE THREAD

**Issue by Noobware1** [ethanblake4/dart_eval#130]
sorry @ethanblake4 for all the dumb issues but i really don't know why can't this code compile. 

```
  Duration parseDuration(String d) {
    final l = d.toLowerCase().split(' ');
    if (l.length > 1) {
      return Duration(
          hours: StringUtils.toInt(StringUtils.substringBefore(l[0], 'h')),
          minutes: StringUtils.toInt(StringUtils.substringBefore(l[1], 'm')));
    } else {
      return Duration(
          minutes: StringUtils.toInt(StringUtils.substringBefore(l[0], 'm')));
    }
  }
```

if i remove the if statement it compiles with ease. The acutal code is more than this but this is function inside a class

here's the full error 
```
RangeError (index): Invalid value: Valid value range is empty: 0
#0      List.[] (dart:core-patch/growable_array.dart:264:36)
#1      Variable.boxIfNeeded (package:dart_eval/src/eval/compiler/variable.dart:73:34)
#2      ScopeContext.resolveBranchStateDiscontinuity.<anonymous closure> (package:dart_eval/src/eval/compiler/context.dart:126:36)
#3      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#4      ScopeContext.resolveBranchStateDiscontinuity (package:dart_eval/src/eval/compiler/context.dart:123:23)
#5      macroBranch (package:dart_eval/src/eval/compiler/macros/branch.dart:23:7)
#6      compileIfStatement (package:dart_eval/src/eval/compiler/statement/if.dart:10:10)
#7      compileStatement (package:dart_eval/src/eval/compiler/statement/statement.dart:38:12)
#8      compileBlock (package:dart_eval/src/eval/compiler/statement/block.dart:16:20)
#9      compileMethodDeclaration (package:dart_eval/src/eval/compiler/declaration/method.dart:54:14)
#10     compileDeclaration (package:dart_eval/src/eval/compiler/declaration/declaration.dart:21:12)
#11     compileClassDeclaration (package:dart_eval/src/eval/compiler/declaration/class.dart:35:5)
#12     compileDeclaration (package:dart_eval/src/eval/compiler/declaration/declaration.dart:17:5)
#13     Compiler.compileSources.<anonymous closure>.<anonymous closure> (package:dart_eval/src/eval/compiler/compiler.dart:427:11)
#14     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#15     Compiler.compileSources.<anonymous closure> (package:dart_eval/src/eval/compiler/compiler.dart:416:15)
#16     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#17     Compiler.compileSources (package:dart_eval/src/eval/compiler/compiler.dart:413:32)
#18     Compiler.compile (package:dart_eval/src/eval/compiler/compiler.dart:146:12)
#19     compilerEval (package:extenstions_test/extenstions_test.dart:78:28)
<asynchronous suspension>
#20     main (package:extenstions_test/extenstions_test.dart:174:20)
```

**Comment by ethanblake4** [comment]
Hi, can you please provide the source for `StringUtils`. I tried with the following code
```dart
Duration parseDuration(String d) {
  final l = d.toLowerCase().split(' ');
  if (l.length > 1) {
    return Duration(
        hours: int.parse(l[0].substring(0, 1)),
        minutes: int.parse(l[1].substring(0, 1)));
  } else {
    return Duration(minutes: int.parse(l[0].substring(0, 1)));
  }
}
```
and did not get an error (however I am using v0.7.0).

**Comment by Noobware1** [comment]
```import 'package:dart_eval/dart_eval_bridge.dart';
import 'package:dart_eval/src/eval/runtime/runtime.dart';
import 'package:dart_eval/stdlib/core.dart';
import 'package:extenstions_test/constants/bridge.dart';
import 'package:extenstions_test/extenstions/string.dart';

class StringUtils {
  static num toNum(String str) => str.toNum();

  static num? toNumOrNull(String str) => str.toNumOrNull();

  static int toInt(String str) => str.toInt();

  static int? toIntOrNull(String str) => str.toIntOrNull();

  static double toDouble(String str) => str.toDouble();

  static double? toDoubleOrNull(String str) => str.toDoubleOrNull();

  static String trimNewLines(String str) => str.trimNewLines();

  // List<int> decodeHex() {
  //   assert(length % 2 == 0, "Must have an even length");
  //   return RegExp('.{1,2}').allMatches(this).map((match) {
  //     return int.parse(match.group(0)!, radix: 16);
  //   }).toList();
  // }

  static String substringBefore(String str, String pattern) =>
      str.substringBefore(pattern);

  static String substringAfter(String str, String pattern) =>
      str.substringAfter(pattern);

  static String substringBeforeLast(String str, String pattern) =>
      str.substringBefore(pattern);

  static String substringAfterLast(String str, String pattern) =>
      str.substringAfterLast(pattern);

  static String toUpperCaseFirst(String str) => str.toUpperCaseFirst();

  static String substringSafe(String str, int start, [int? end]) =>
      str.substringSafe(start, end);

  static String trim(String str) => str.trim();

  static String trimAllSides(String str) => str.trim().trimLeft().trimRight();
}

class $StringUtils implements $Instance {
  $StringUtils.wrap(this.$value);

  static void configureForRuntime(Runtime runtime) {
    runtime.registerBridgeFunc(bridgeLibary, 'StringUtils.', $new);
    runtime.registerBridgeFunc(bridgeLibary, 'StringUtils.toNum', $toNum);

    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.toNumOrNull', $toNumOrNull);

    runtime.registerBridgeFunc(bridgeLibary, 'StringUtils.toInt', $toInt);

    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.toIntOrNull', $toIntOrNull);

    runtime.registerBridgeFunc(bridgeLibary, 'StringUtils.toDouble', $toDouble);

    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.toDoubleOrNull', $toDoubleOrNull);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.substringBefore', $substringBefore);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.substringAfter', $substringAfter);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.substringBeforeLast', $substringBeforeLast);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.substringAfterLast', $substringAfterLast);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.toUpperCaseFirst', $toUpperCaseFirst);

    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.trimNewLines', $trimNewLines);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.substringSafe', $substringSafe);

    runtime.registerBridgeFunc(bridgeLibary, 'StringUtils.trim', $trim);
    runtime.registerBridgeFunc(
        bridgeLibary, 'StringUtils.trimAllSides', $trimAllSides);
  }

  static const $type =
      BridgeTypeRef(BridgeTypeSpec(bridgeLibary, 'StringUtils'));

  static const $param = BridgeParameter('str',
      BridgeTypeAnnotation(BridgeTypeRef.type(RuntimeTypes.stringType)), false);

  static const $patterParam = BridgeParameter('pattern',
      BridgeTypeAnnotation(BridgeTypeRef.type(RuntimeTypes.stringType)), false);

  static const $declaration = BridgeClassDef(BridgeClassType($type),
      constructors: {
        '': BridgeConstructorDef(BridgeFunctionDef(
            returns: BridgeTypeAnnotation($type), namedParams: [], params: []))
      },
      methods: {
        'toNum': BridgeMethodDef(
            BridgeFunctionDef(
                returns: BridgeTypeAnnotation(
                    BridgeTypeRef.type(RuntimeTypes.numType)),
                params: [$param]),
            isStatic: true),
        'toNumOrNull': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.numType),
                  nullable: true),
              params: [$param],
            ),
            isStatic: true),
        'toInt': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.intType)),
              params: [$param],
            ),
            isStatic: true),
        'toIntOrNull': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.intType),
                  nullable: true),
              params: [$param],
            ),
            isStatic: true),
        'toDouble': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.doubleType)),
              params: [$param],
            ),
            isStatic: true),
        'toDoubleOrNull': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.doubleType),
                  nullable: true),
              params: [$param],
            ),
            isStatic: true),
        'trimNewLines': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param],
            ),
            isStatic: true),
        'substringBefore': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param, $patterParam],
            ),
            isStatic: true),
        'substringAfter': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param, $patterParam],
            ),
            isStatic: true),
        'substringBeforeLast': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param, $patterParam],
            ),
            isStatic: true),
        'substringAfterLast': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param, $patterParam],
            ),
            isStatic: true),
        'toUpperCaseFirst': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param],
            ),
            isStatic: true),
        'substringSafe': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [
                $param,
                BridgeParameter(
                    'start',
                    BridgeTypeAnnotation(
                        BridgeTypeRef.type(RuntimeTypes.intType)),
                    false),
                BridgeParameter(
                    'end',
                    BridgeTypeAnnotation(
                        BridgeTypeRef.type(RuntimeTypes.intType),
                        nullable: true),
                    true),
              ],
            ),
            isStatic: true),
        'trim': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param],
            ),
            isStatic: true),
        'trimAllSides': BridgeMethodDef(
            BridgeFunctionDef(
              returns: BridgeTypeAnnotation(
                  BridgeTypeRef.type(RuntimeTypes.stringType)),
              params: [$param],
            ),
            isStatic: true),
      },
      fields: {},
      getters: {},
      setters: {},
      wrap: true);

  @override
  $Value? $getProperty(Runtime runtime, String identifier) {
    return null;
  }

  static $Value? $new(Runtime runtime, $Value? target, List<$Value?> args) {
    return $StringUtils.wrap(StringUtils());
  }

  static $Value? $toNum(Runtime runtime, $Value? target, List<$Value?> args) {
    return $num(StringUtils.toNum(args[0]?.$value));
  }

  static $Value? $toNumOrNull(
      Runtime runtime, $Value? target, List<$Value?> args) {
    final value = StringUtils.toNumOrNull(args[0]?.$value);
    if (value == null) return null;
    return $num(value);
  }

  static $Value? $toInt(Runtime runtime, $Value? target, List<$Value?> args) {
    return $int(StringUtils.toInt(args[0]?.$value));
  }

  static $Value? $toIntOrNull(
      Runtime runtime, $Value? target, List<$Value?> args) {
    final value = StringUtils.toIntOrNull(args[0]?.$value);
    return value == null ? null : $int(value);
  }

  static $Value? $toDouble(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $double(StringUtils.toDouble(args[0]?.$value));
  }

  static $Value? $toDoubleOrNull(
      Runtime runtime, $Value? target, List<$Value?> args) {
    final value = StringUtils.toDoubleOrNull(args[0]?.$value);
    return value == null ? null : $double(value);
  }

  static $Value? $trimNewLines(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(StringUtils.trimNewLines(args[0]?.$value));
  }

  static $Value? $substringBefore(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(
        StringUtils.substringBefore(args[0]?.$value, args[1]?.$value));
  }

  static $Value? $substringAfter(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(
        StringUtils.substringAfter(args[0]?.$value, args[1]?.$value));
  }

  static $Value? $substringBeforeLast(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(
        StringUtils.substringAfterLast(args[0]?.$value, args[1]?.$value));
  }

  static $Value? $substringAfterLast(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(
        StringUtils.substringAfterLast(args[0]?.$value, args[1]?.$value));
  }

  static $Value? $toUpperCaseFirst(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(StringUtils.toUpperCaseFirst(args[0]?.$value));
  }

  static $Value? $substringSafe(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(StringUtils.substringSafe(
        args[0]?.$value, args[1]?.$value, args[2]?.$value));
  }

  static $Value? $trim(Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(StringUtils.trim(args[0]?.$value));
  }

  static $Value? $trimAllSides(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $String(StringUtils.trimAllSides(args[0]?.$value));
  }

  @override
  int $getRuntimeType(Runtime runtime) => runtime.lookupType($type.spec!);

  @override
  get $reified => $value;

  @override
  void $setProperty(Runtime runtime, String identifier, $Value value) {
    // TODO: implement $setProperty
  }

  @override
  final StringUtils $value;
}
```

**Comment by Noobware1** [comment]
I don't think string utils the problem because they work in my other code just fine

**Comment by ethanblake4** [comment]
Appears to be fixed in v0.7.0
